src/middlewared/middlewared/plugins/auth.py

```diff
@@ -291,13 +295,23 @@ def get_token(self, token_id):
             return None
 
     @no_auth_required
-    @accepts(Str('username'), Str('password'))
+    @accepts(Str('username'), Str('password'), Str('otp_token', null=True, default=None))
     @pass_app
-    async def login(self, app, username, password):
-        """Authenticate session using username and password.
+    async def login(self, app, username, password, otp_token=None):
+        """
+        Authenticate session using username and password.
         Currently only root user is allowed.
+        `otp_token` must be specified if two factor authentication is enabled.
         """
         valid = await self.check_user(username, password)
+        twofactor_auth = await self.middleware.call('auth.twofactor.config')
+
+        if twofactor_auth['enabled']:
+            valid &= await self.middleware.call(
```

Comment:
Please leave a comment here about why we are doing this even if `check_user` returns false so it won't get "optimized" by someone one day.